# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [1]:
# TODO: import the pandas library
import pandas as pd

# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)

# TODO: remove the 'Unnamed:62' column from each data set
df_rural.drop('Unnamed: 62', axis=1, inplace=True)
df_electricity.drop('Unnamed: 62', axis=1, inplace=True)

# TODO: combine the two data sets together using the concat method
# In other words, all of the words of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df = pd.concat([df_rural, df_electricity])

In [2]:
df_rural.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  ...  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300  ...   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250  ...   
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...   
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770  ...   
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555  ...   

     2008    2009    2010    2011    2012    2013    2014    2015    2016  \
0  56.217  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696   
1  76.054  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868   
2  61.491  60.701  59.903  59.100  58.301  57.510  56.726  55.950  55.181   
3  50.009  48.924  47.837  46.753  45.670  44.617  43.591  42.593  41.624   
4  11.133  11.648  12.183  12.740  13.292  13.835  14.367  14.885  15.388   

     2017  
0  58.893  
1  72.430  
2  54.422  
3  40.684  
4  15.873  

[5 rows x 62 columns]

In [3]:
df_electricity.head()

Country Name Country Code                           Indicator Name  \
0        Aruba          ABW  Access to electricity (% of population)   
1  Afghanistan          AFG  Access to electricity (% of population)   
2       Angola          AGO  Access to electricity (% of population)   
3      Albania          ALB  Access to electricity (% of population)   
4      Andorra          AND  Access to electricity (% of population)   

   Indicator Code  1960  1961  1962  1963  1964  1965  ...        2008  \
0  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   93.086166   
1  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   42.400000   
2  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   31.268013   
3  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  100.000000   
4  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  100.000000   

         2009        2010        2011        2012        2013        2014  \
0   93.354546   93.356292   93.942375   94.255814   94.578262   94.906723   
1   44.854885   42.700000   43.222019   69.100000   67.259552   89.500000   
2   32.382469   33.514950   34.600000   35.821964   36.990490   32.000000   
3  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2015        2016  2017  
0   95.238182   95.570145   NaN  
1   71.500000   84.137138   NaN  
2   42.000000   40.520607   NaN  
3  100.000000  100.000000   NaN  
4  100.000000  100.000000   NaN  

[5 rows x 62 columns]

# Exercise 2 (Challenge)

This exercise is more challenging. Combine the data in the csv file together to the output that look like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [4]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value

df_rural_melt = pd.melt(df_rural,\
                        id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],\
                       var_name = 'Year', value_name='Electricity_Value')
df_electricity_melt = pd.melt(df_electricity,\
                              id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],\
                             var_name='Year', value_name='Rural_Value')

# TODO: drop any columns from the data frames that aren't needed
df_rural_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)
df_electricity_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_merge = df_rural_melt.merge(df_electricity_melt, how='outer',\
                               on=['Country Name', 'Country Code', 'Year'])

# TODO: sort the results by country and then by year
df_combined = df_merge.sort_values(['Country Name', 'Year'])
df_combined

Country Name Country Code  Year  Electricity_Value  Rural_Value
1      Afghanistan          AFG  1960             91.779          NaN
265    Afghanistan          AFG  1961             91.492          NaN
529    Afghanistan          AFG  1962             91.195          NaN
793    Afghanistan          AFG  1963             90.890          NaN
1057   Afghanistan          AFG  1964             90.574          NaN
1321   Afghanistan          AFG  1965             90.250          NaN
1585   Afghanistan          AFG  1966             89.915          NaN
1849   Afghanistan          AFG  1967             89.570          NaN
2113   Afghanistan          AFG  1968             89.214          NaN
2377   Afghanistan          AFG  1969             88.848          NaN
2641   Afghanistan          AFG  1970             88.471          NaN
2905   Afghanistan          AFG  1971             88.083          NaN
3169   Afghanistan          AFG  1972             87.684          NaN
3433   Afghanistan          AFG  1973             87.274          NaN
3697   Afghanistan          AFG  1974             86.851          NaN
3961   Afghanistan          AFG  1975             86.417          NaN
4225   Afghanistan          AFG  1976             85.971          NaN
4489   Afghanistan          AFG  1977             85.513          NaN
4753   Afghanistan          AFG  1978             85.042          NaN
5017   Afghanistan          AFG  1979             84.565          NaN
5281   Afghanistan          AFG  1980             84.319          NaN
5545   Afghanistan          AFG  1981             84.070          NaN
5809   Afghanistan          AFG  1982             83.818          NaN
6073   Afghanistan          AFG  1983             83.563          NaN
6337   Afghanistan          AFG  1984             83.304          NaN
6601   Afghanistan          AFG  1985             83.042          NaN
6865   Afghanistan          AFG  1986             82.777          NaN
7129   Afghanistan          AFG  1987             82.509          NaN
7393   Afghanistan          AFG  1988             82.237          NaN
7657   Afghanistan          AFG  1989             81.962          NaN
...            ...          ...   ...                ...          ...
7655      Zimbabwe          ZWE  1988             72.476          NaN
7919      Zimbabwe          ZWE  1989             71.750          NaN
8183      Zimbabwe          ZWE  1990             71.012    30.357832
8447      Zimbabwe          ZWE  1991             70.262    30.718805
8711      Zimbabwe          ZWE  1992             69.501    28.200000
8975      Zimbabwe          ZWE  1993             69.060    31.436678
9239      Zimbabwe          ZWE  1994             68.665    28.100000
9503      Zimbabwe          ZWE  1995             68.268    32.130077
9767      Zimbabwe          ZWE  1996             67.868    32.459949
10031     Zimbabwe          ZWE  1997             67.466    32.774529
10295     Zimbabwe          ZWE  1998             67.061    33.070751
10559     Zimbabwe          ZWE  1999             66.653    38.400000
10823     Zimbabwe          ZWE  2000             66.242    33.597420
11087     Zimbabwe          ZWE  2001             65.830    33.830868
11351     Zimbabwe          ZWE  2002             65.415    34.200000
11615     Zimbabwe          ZWE  2003             65.521    34.266785
11879     Zimbabwe          ZWE  2004             65.706    34.481380
12143     Zimbabwe          ZWE  2005             65.890    34.701820
12407     Zimbabwe          ZWE  2006             66.074    37.200000
12671     Zimbabwe          ZWE  2007             66.257    35.184479
12935     Zimbabwe          ZWE  2008             66.440    35.457310
13199     Zimbabwe          ZWE  2009             66.622    43.369082
13463     Zimbabwe          ZWE  2010             66.804    36.063057
13727     Zimbabwe          ZWE  2011             66.985    36.900000
13991     Zimbabwe          ZWE  2012             67.166    36.728878
14255     Zimbab

In [5]:
df_rural.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)
df_rural_melt = pd.melt(df_rural,\
                        id_vars=['Country Name', 'Country Code'],\
                       var_name = 'Year', value_name='Electricity_Value')
df_rural_melt.head()

Country Name Country Code  Year  Electricity_Value
0        Aruba          ABW  1960             49.224
1  Afghanistan          AFG  1960             91.779
2       Angola          AGO  1960             89.565
3      Albania          ALB  1960             69.295
4      Andorra          AND  1960             41.550

In [7]:
df_combined[df_combined['Country Code']=='ABW']

Country Name Country Code  Year  Electricity_Value  Rural_Value
0            Aruba          ABW  1960             49.224          NaN
264          Aruba          ABW  1961             49.239          NaN
528          Aruba          ABW  1962             49.254          NaN
792          Aruba          ABW  1963             49.270          NaN
1056         Aruba          ABW  1964             49.285          NaN
1320         Aruba          ABW  1965             49.300          NaN
1584         Aruba          ABW  1966             49.315          NaN
1848         Aruba          ABW  1967             49.330          NaN
2112         Aruba          ABW  1968             49.346          NaN
2376         Aruba          ABW  1969             49.361          NaN
2640         Aruba          ABW  1970             49.376          NaN
2904         Aruba          ABW  1971             49.391          NaN
3168         Aruba          ABW  1972             49.407          NaN
3432         Aruba          ABW  1973             49.422          NaN
3696         Aruba          ABW  1974             49.437          NaN
3960         Aruba          ABW  1975             49.452          NaN
4224         Aruba          ABW  1976             49.468          NaN
4488         Aruba          ABW  1977             49.483          NaN
4752         Aruba          ABW  1978             49.498          NaN
5016         Aruba          ABW  1979             49.513          NaN
5280         Aruba          ABW  1980             49.528          NaN
5544         Aruba          ABW  1981             49.544          NaN
5808         Aruba          ABW  1982             49.559          NaN
6072         Aruba          ABW  1983             49.574          NaN
6336         Aruba          ABW  1984             49.589          NaN
6600         Aruba          ABW  1985             49.605          NaN
6864         Aruba          ABW  1986             49.620          NaN
7128         Aruba          ABW  1987             49.635          NaN
7392         Aruba          ABW  1988             49.650          NaN
7656         Aruba          ABW  1989             49.665          NaN
7920         Aruba          ABW  1990             49.681    88.445351
8184         Aruba          ABW  1991             49.696    88.780846
8448         Aruba          ABW  1992             50.002    89.115829
8712         Aruba          ABW  1993             50.412    89.447754
8976         Aruba          ABW  1994             50.823    89.773560
9240         Aruba          ABW  1995             51.233    90.090187
9504         Aruba          ABW  1996             51.644    90.394585
9768         Aruba          ABW  1997             52.054    90.683678
10032        Aruba          ABW  1998             52.464    90.954422
10296        Aruba          ABW  1999             52.873    91.203751
10560        Aruba          ABW  2000             53.283    91.660398
10824        Aruba          ABW  2001             53.661    91.638092
11088        Aruba          ABW  2002             54.028    91.833717
11352        Aruba          ABW  2003             54.394    92.023048
11616        Aruba          ABW  2004             54.760    92.212166
11880        Aruba          ABW  2005             55.125    92.407120
12144        Aruba          ABW  2006             55.489    92.613983
12408        Aruba          ABW  2007             55.853    92.838821
12672        Aruba          ABW  2008             56.217    93.086166
12936        Aruba          ABW  2009             56.579    93.354546
13200        Aruba          ABW  2010             56.941    93.356292
13464        Aruba          ABW  2011             57.302    93.942375
13728        Aruba          ABW  2012             57.636    94.255814
13992        Aruba          ABW  2013             57.942    94.578262
14256        Aruba          ABW  2014             58.221    94.906723
14520        Aruba          ABW  2015             58.472    95.238182
14784        Aru